File: Olympiad/POI/official/2013-XX/gra/prog/gra2.cpp
/*************************************************************************
 *                                                                       *
 *                    XX Olimpiada Informatyczna                         *
 *                                                                       *
 *   Zadanie:              Gra Tower Defense                             *
 *   Autor:                Lech Duraj                                    *
 *   Zlozonosc czasowa:    O(n+m)                                        *
 *   Zlozonosc pamieciowa: O(n+m)                                        *
 *   Opis:                 Rozwiazanie wzorcowe                          *
 *                                                                       *
 *************************************************************************/

#include <cstdio>
#include <algorithm>
#include <vector>

using namespace std;

#define foreach(t,v) for(typeof(v.begin()) t = v.begin(); t!=v.end(); t++)

const int maxn = 500*1000;


int main()
{
    int n,m,k;
    vector<int> G[maxn+1];
    bool marked[maxn+1];
    scanf("%d %d %d",&n,&m,&k);
    for(int i=0; i<m; i++)
    {
        int a,b;
        scanf("%d %d",&a,&b);
        G[a].push_back(b);
        G[b].push_back(a);
    }
    for(int i=1; i<=n; i++)
        marked[i] = false;
    vector<int> towers;
    for(int i=1; i<=n; i++)
        if (!marked[i])
        {
            towers.push_back(i);
            marked[i] = true;
            foreach(t,G[i])
            {
                marked[*t] = true;
                foreach(s,G[*t])
                marked[*s] = true;
            }
        }
    printf("%d\n",(int)towers.size());
    foreach(t,towers)
    printf("%d ",*t);
}

File: Olympiad/POI/official/2010-XVII/gra/prog/gra2.pas
(*************************************************************************
 *                                                                       *
 *                    XVII Olimpiada Informatyczna                       *
 *                                                                       *
 *   Zadanie:           Gra w minima (GRA)                               *
 *   Plik:              gra2.pas                                         *
 *   Autor:             Jakub Lacki                                      *
 *   Opis:              Rozwiazanie wzorcowe.                            *
 *   Zlozonosc czasowa: O(n * log n)                                     *
 *                                                                       *
 *************************************************************************)

const MAXN = 1000000;

type tab = array [0..MAXN-1] of longint;

procedure quicksort (var t:tab; l,r:longint);
   var 
       p,tmp:int64;
       i,j:longint;
   begin 
       if l < r then
       begin
           p := t[random(r-l) + l+1];
           i := l-1;
           j := r+1;
           repeat
               repeat i := i+1 until p <= t[i];
               repeat j := j-1 until p >= t[j];

               tmp:=t[i]; t[i]:=t[j]; t[j]:=tmp

           until i >= j;

           t[j]:=t[i];
	   t[i]:=tmp;

           quicksort(t,l,i-1);
           quicksort(t,i,r)
       end
   end;


var
    karty : array [0..MAXN-1] of longint;
    wynik : array [0..MAXN-1] of int64;
    n, i : longint;

begin
    read(n);
    for i:=0 to n-1 do
	read(karty[i]);

    quicksort(karty, 0, n-1);

    wynik[0] := karty[0];

    for i:=1 to n-1 do
	if wynik[i-1] > karty[i] - wynik[i-1] then
	    wynik[i] := wynik[i-1]
	else 
    	    wynik[i] := karty[i] - wynik[i-1];

    writeln(wynik[n-1]);
end.
